lib/Encode/ZapCP1252.pm

```diff
@@ -73,40 +73,135 @@ our %utf8_for = (
+# have the characteristic that the first two bits are both one.  The number of
+# bytes that form a character corresponds to the number of consecutive leading
+# bits that are all one in the start byte.  In the case of FE, the first 7
+# bits are one, so the number of bytes in the character it represents is 7.
+# FF is a special case, and Perl has arbitrarily set it to 13 instead of the
+# expected 8.
+#
+# The remaining bytes begin with '10', from 80..9F.  They are called
+# continuation bytes, and a UTF-8 character is comprised of a start byte
+# indicating 'n' bytes total in it, then 'n-1' of these continuation bytes.
+# What the character is that each sequence represents is derived by shifting
+# and adding the other bits in the bytes.  (C0 and C1 aren't actually legal
+# start bytes for security reasons that need not concern us here, hence are
+# marked as negative in the table below.)
+
+  # 0  1  2  3  4  5  6  7  8  9  A  B  C  D  E  F
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  # 0
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  # 1
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  # 2
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  # 3
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  # 4
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  # 5
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  # 6
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,  # 7
+   -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,  # 8
+   -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,  # 9
+   -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,  # A
+   -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,  # B
+   -1,-1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,  # C
+    2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,  # D
+    3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,  # E
+    4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 6, 6, 7,13,  # F
+);
+
 BEGIN {
     my $proto = $] >= 5.010000 ? '_' : '$';
-    eval "sub zap_cp1252($proto) { &_ascii; }";
-    eval "sub fix_cp1252($proto) { &_utf8; }";
+    eval "sub zap_cp1252($proto) { unshift \@_, \\%ascii_for; &_tweakit; }";
+    eval "sub fix_cp1252($proto) { unshift \@_, \\%utf8_for;  &_tweakit; }";
 }
 
-sub _ascii {
+# These are the bytes that CP1252 redefines
+my $cp1252_re = qr/[\x80\x82-\x8c\x8e\x91-\x9c\x9e\x9f]/;
+
+sub _tweakit {
+    my $table = shift;
     return unless defined $_[0];
     local $_[0] = $_[0] if defined wantarray;
-    if (PERL588 && Encode::is_utf8($_[0]) && !utf8::valid($_[0])) {
+    my $is_utf8 = PERL588 && Encode::is_utf8($_[0]);
+    my $valid_utf8 = $is_utf8 && utf8::valid($_[0]);
+    if (!$is_utf8) {
+
+        # Here is non-UTF-8. Change the 1252 characters to their UTF-8
+        # counterparts. (These bytes are very rarely used in real world
+        # applications, so their presence likely indicates that CP1252 was
+        # meant, regardless of whether the string is UTF-8 or not.)
+        $_[0] =~ s/($cp1252_re)/$table->{$1}/gems;
```

Comment:
I don't understand.  One of your commits wrapped $table with encode_utf8(), and it created a compilation error.  But I think that was the right thing to do, and I don't see how it got removed, unless github did it manually

Reply:
I removed `utf8` and switched things to call `decode_utf8()` where appropriate, instead, because the module is mainly intended for dealing with bytes, with Unicode string handling as a secondary, less recommended use case.